Use DeleteByIndex in OutboxRecordsCleaner, execute with Task.Delay loop

DavidBoike opened the pull request:
Connects to #255 

Will probably have to backport (manually) to 3.0.x as well.

Please review @Particular/ravendb-persistence-maintainers

DavidBoike commented on src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs:
```diff
@@ -30,59 +31,79 @@ protected override void Setup(FeatureConfigurationContext context)
+                    }
+                    catch (Exception ex)
+                    {
+                        logger.Warn("Unable to remove expired Outbox records from Raven database.", ex);
```
Should we also have a circuit breaker here?

mauroservienti replied:
you mean: shut down the endpoint if for whatever reason we cannot clean up records for more than x attempts?

DavidBoike replied:
Perhaps, or at the very least log ERROR instead of WARN?

At some point this little problem will escalate into a really big problem.

mauroservienti replied:
:+1:  let's start easy: change from WARN to ERROR

johnsimons commented:
@DavidBoike the usage of `IndexQuery` I would probably include also, `PageSize` (otherwise defaults to 128), `Cutoff = SystemTime.UtcNow` and `DisableCaching = true`, we use this in ServiceControl

johnsimons commented:
In ServiceControl we changed our date range queries to use Ticks instead, so the comparisons are all numeric which is faster, this would mean you would have to change your index to index ticks instead of datetime, but in the end you get faster queries.
Eg. [here](https://github.com/Particular/ServiceControl/blob/45d6c5c0dad9334fa44928cb77a21a4bd9e98623/src/ServiceControl/Infrastructure/RavenDB/Expiration/AuditMessageCleaner.cs#L31) and [here](https://github.com/Particular/ServiceControl/blob/45d6c5c0dad9334fa44928cb77a21a4bd9e98623/src/ServiceControl/Infrastructure/RavenDB/Expiration/ExpiryProcessedMessageIndex.cs#L14)

DavidBoike commented:
@johnsimons: Raven only uses PageSize for queries. It [doesn't honor that when doing DeleteByIndex](https://groups.google.com/d/msg/ravendb/yrKTDt6UWx4/WF5KfjtiNssJ) - it deletes every matching document, in excess even of 1024. Similarly, DisableCaching is only for queries. As for CutOff, that's already covered by AllowStale = true. Especially since we're deleting 7-day-old data (default window) we don't really care what the status of the index is, we'll just delete everything that the index says is older than 7 days right now.

Using Ticks would be an interesting optimization, but I'm not sure that change is worth the effort. That would require a complete reindex, and then would create a problem if the customer used the same database for multiple endpoints and didn't upgrade them all in lockstep. Or it would require a completely new index, which wouldn't be small, and the customer would have to manage removing the old one. Other thoughts on that?

johnsimons commented:
> Raven only uses PageSize for queries. It doesn't honor that when doing DeleteByIndex - it deletes every matching document, in excess even of 1024. Similarly, DisableCaching is only for queries. As for CutOff, that's already covered by AllowStale = true. Especially since we're deleting 7-day-old data (default window) we don't really care what the status of the index is, we'll just delete everything that the index says is older than 7 days right now.

nice, I never know so I just add everything just to make sure


> Using Ticks would be an interesting optimization, but I'm not sure that change is worth the effort.

If performance is an issue, then it is worth it, deletes in raven affect the inserts performance, so I would recommend testing it with a loaded system (eg 1500msg/sec)

mauroservienti commented on src/NServiceBus.RavenDB/Outbox/OutboxRecordsCleaner.cs:
```diff
@@ -15,61 +13,26 @@ public OutboxRecordsCleaner(IDocumentStore documentStore)
-                             * found in the streamed document metadata.
-                             */
-                            var key = enumerator.Current.Key ?? enumerator.Current.Metadata["@id"].ToString();
+            var operation = await documentStore.AsyncDatabaseCommands.DeleteByIndexAsync(nameof(OutboxRecordsIndex), query, bulkOpts, cancellationToken)
```
`nameof` here might be dangerous since we are starting from the assumption that we won't change/customize the index name in any way. E.g. if the call will be named `Outbox_RecordsIndex` the index name will be `Outbox/RecordsIndex` in RavenDB.

DavidBoike commented:
* Made the Lucene query `$"Dispatched:true AND DispatchedAt:[* TO {dateTime:o}]"` because non-dispatched records were not affected simply because of an accident that "null" sorts alphabetically after numbers! Including `Dispatched:true` makes it explicit.
* Referred to the index by string name as per @mauroservienti's suggestion
* Changed the log level when we can't clean the Outbox to `ERROR`. If we get reports of this being a problem, we can introduce a circuit breaker.
* Removed `MessageId` from the OutboxRecordsIndex. It's not needed for anything so we are wastefully storing a stringified Guid.

johnsimons commented:
@DavidBoike regarding the index name, the actual code you should do is:
```
var indexName = new OutboxRecordsIndex().IndexName;
```

mauroservienti commented:
Reviewed with @weralabaj, @DavidBoike we've incorporated @johnsimons's suggestion, feel free to merge. Looks good.
